server/src/test/java/org/elasticsearch/health/node/ShardsCapacityHealthIndicatorServiceTests.java

```diff
@@ -0,0 +1,407 @@
+        var randomMaxShardsPerNodeSetting = randomInt();
+        Function<Integer, ShardsCapacityHealthIndicatorService.ShardsCapacityChecker> checkerWrapper = shardsToAdd -> (
+            maxConfiguredShardsPerNode,
+            numberOfNewShards,
+            replicas,
+            state) -> {
+            assertEquals(mockedState, state);
+            assertEquals(randomMaxShardsPerNodeSetting, maxConfiguredShardsPerNode);
+            return new ShardLimitValidator.Result(
+                numberOfNewShards != shardsToAdd && replicas == 1,
+                Optional.empty(),
+                randomInt(),
+                randomInt(),
+                randomAlphaOfLength(5)
+            );
+        };
+
+        assertEquals(calculateFrom(randomMaxShardsPerNodeSetting, mockedState, checkerWrapper.apply(5)).status(), RED);
+        assertEquals(calculateFrom(randomMaxShardsPerNodeSetting, mockedState, checkerWrapper.apply(10)).status(), YELLOW);
+
+        // Let's cover the holes :)
+        Stream.of(randomIntBetween(1, 4), randomIntBetween(6, 9), randomIntBetween(11, Integer.MAX_VALUE))
+            .map(checkerWrapper)
+            .map(checker -> calculateFrom(randomMaxShardsPerNodeSetting, mockedState, checker))
+            .map(ShardsCapacityHealthIndicatorService.StatusResult::status)
+            .forEach(status -> assertEquals(status, GREEN));
+    }
+
+    // We expose the indicator name and the diagnoses in the x-pack usage API. In order to index them properly in a telemetry index
+    // they need to be declared in the health-api-indexer.edn in the telemetry repository.
+    public void testMappedFieldsForTelemetry() {
+        assertEquals(ShardsCapacityHealthIndicatorService.NAME, "shards_capacity");
+        assertEquals(
+            "elasticsearch:health:shards_capacity:diagnosis:increase_max_shards_per_node",
+            SHARDS_MAX_CAPACITY_REACHED_DATA_NODES.definition().getUniqueId()
+        );
+        assertEquals(
+            "elasticsearch:health:shards_capacity:diagnosis:increase_max_shards_per_node_frozen",
+            SHARDS_MAX_CAPACITY_REACHED_FROZEN_NODES.definition().getUniqueId()
+        );
+    }
+
+    private static int randomValidMaxShards() {
+        return randomIntBetween(50, 1000);
+    }
+
+    private Map<String, Object> xContentToMap(ToXContent xcontent) throws IOException {
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        xcontent.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        XContentParser parser = XContentType.JSON.xContent()
+            .createParser(xContentRegistry(), LoggingDeprecationHandler.INSTANCE, BytesReference.bytes(builder).streamInput());
+        return parser.map();
+    }
+
+    private static ClusterService createClusterService(ClusterState clusterState) {
+        var clusterService = mock(ClusterService.class);
+        when(clusterService.state()).thenReturn(clusterState);
+        return clusterService;
+    }
```

Comment:
We prefer to not mock things when we can avoid it (so the test is as close to the real scenario as possible).

Please replace this mock with a real `ClusterService`. 
See
```
org.elasticsearch.test.ClusterServiceUtils.createClusterService
```
and
```
org.elasticsearch.test.ClusterServiceUtils.setState
```

Reply:
real nice! I did copy what we're using here https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/health/node/DiskHealthIndicatorServiceTests.java#L1065-L1073